lib/palletjack/ipa.rb

```diff
@@ -0,0 +1,273 @@
+
+        @@ipa_url = "https://" + target_server + "/ipa"
+      end
+
+      def _make_request # :nodoc:
+
+        # Authenticate to request a new session
+        # By creating a session and then using the session ID, we do not need
+        # to do a GSSAPI Authentication for every API request made within the
+        # same script.
+        if @@session_init == false then
+          # The session is automatically initialized if a request is made towards the API
+          _login_payload = {
+                "id"     => 0, 
+                "method" => "ping",
+                "params" => [ [], {} ]
+          }
+
+          # This should act as default settings troughout the session
+          @@curl.set(:HTTPAUTH, Curl::CURLAUTH_GSSNEGOTIATE)
+          @@curl.username = ':'
+          @@curl.enable_cookies = true
+          @@curl.verbose = false
+          @@curl.headers["Referer"] = @@ipa_url + "/json"
+          @@curl.headers['Content-Type'] = "application/json"
+          @@curl.url = @@ipa_url + "/json"
+          @@curl.http_post(_login_payload.to_json)
+          _session_result = JSON.parse(@@curl.body_str)
+
+          if _session_result['error'] != nil then
+            puts "Error authentication!"
+          end
+          @@curl.url = @@ipa_url + "/session/json"
+          #@@curl.set(:HTTPAUTH, Curl::CURLAUTH_NONE) # Reset auth, rely on cookie
+        end
+
+        if @debug then
+          puts "DEBUG: Request: " + @payload
+        end
+
+        @@curl.http_post(@payload)
+
+        @body = @@curl.body_str
+
+        if @debug then
+          puts "DEBUG: Response:" + @body
+        end
+
+      end
+
+      # Return a +hash+ with the command response
+      def response
+        JSON.parse(@body)
+      end
+
+      # Return a +string+ with the command reponse JSON.
+      def raw_response
+        @body
+      end
+
+    end
+
+    class Command
+
+      @@api_version = "2.156"
+
+      # Execute RPC Commands on a IPA server.
+      #
+      # :call-seq:
+      #   new(method [,pos1 [,params]]) -> command
+      #
+      # This class will take a method name, positional parameters and named
+      # parameters and make a JSON-RPC request against a IPA server.
+      #
+      # <tt>ipa -vv <command> [params]</tt> may be used to see an example of the data
+      # structure used by the API. In general the API parameters is very similar
+      # to the command line tool.
+      #
+      # === Attributes
+      #
+      # * +method+ - The JSON-RPC Method to call
+      # * +name+ - The positional parameters to the method. Can be an array or string.
+      # * +params+ - Named parameters to the method.
+      # * +debug+ - Enable debugging output, default: false
+      #
+      # Examples:
+      #   <tt>PalletJack::Ipa::Command.new("host_add", "new_system", { ip_address: "192.168.13.37" })</tt>
+      #   <tt>PalletJack::Ipa::Command.new("host_find")</tt>
+      def initialize(method, name=nil, params={}, debug = false)
```

Comment:
`def initialize(method, *args, **kwargs)`
will automatically collect all leftover positional arguments into `args`, and all keyword arguments into `kwargs`